### DISCUSSION TITLE
Breaking change: limit user to import internal module by add exports to package.json

### DISCUSSION THREAD

**Discussion by hardfist** [web-infra-dev/rspack#3963] [Core]
since rspack have lots of internal module, if user use internal module directly then every internal module refactor has breaking change risk, so it would be better to limit user not import internal module directly.

if user want to use some internal module, they should use either from entry or by rspack.[InternalModule]

impacts: 
- [x]  modernjs: modernjs don't use any rspack internal  module right now, so it has no impact
- [ ] nx: nx ecosystem-ci use `@rspack/core/dist/watching` which is a breaking , we should update its usage

**Comment by hardfist** [comment]
cc @mandarini @alexander-akait

**Comment by mandarini** [comment]
@hardfist sorry just saw this, I was on vacation. I'll take a look today or tomorrow!
